src/OrchardCore.Modules/OrchardCore.Media/Processing/AsyncKeyLock.cs

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) Six Labors and contributors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace OrchardCore.Media.Processing
+{
+    /// <summary>
+    /// The async key lock prevents multiple asynchronous threads acting upon the same object with the given key at the same time.
+    /// It is designed so that it does not block unique requests allowing a high throughput.
+    /// </summary>
+    internal sealed class AsyncKeyLock
```

Comment:
I implemented it in ImageSharp, and since then I think someone found an issue and they reverted the code. I will need to look back at it to see if it's really necessary. When I did this change in IS I commented that I thought we didn't need any lock actually.